### PR TITLE
Tests for HandledNodesTasks

### PR DESCRIPTION
**Pull request by dvdplm**
Based on #546

### CODE REVIEW

**Comment by dvdplm** [comment]
Merged in master.

**Comment by tomaka** [comment]
Please remove the printlns, the diff is really verbose.

**Comment by dvdplm** [comment]
Done.

File: core/src/nodes/handled_node.rs [hunk 1]
```diff
@@ -296,6 +298,13 @@ where
                 }
             }
         }
+        // REVIEW: I don't think this can ever happen. The only way to break out
```
Comment (tomaka) [inline core/src/nodes/handled_node.rs]: This is no longer true after #582

File: core/src/nodes/handled_node_tasks.rs [hunk 2]
```diff
@@ -71,6 +72,7 @@ pub struct HandledNodesTasks<TInEvent, TOutEvent, THandler> {
 
 impl<TInEvent, TOutEvent, THandler> fmt::Debug for HandledNodesTasks<TInEvent, TOutEvent, THandler> {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        f.debug_struct("HandledNodesTask").finish().expect("adding a string works");
```
Comment (tomaka) [inline core/src/nodes/handled_node_tasks.rs]: That's not how you do this. The two entries have to be merged into one.
However I don't like this change by itself. This struct is just a collection of tasks and thus can be printed the same way as a collection.

File: core/src/nodes/handled_node_tasks.rs [hunk 3]
```diff
@@ -419,6 +419,7 @@ where
                     // Process the node.
                     loop {
                         match node.poll() {
+                            // REVIEW: I don't think this can happen, see comment in handled_node.rs
```
Comment (tomaka) [inline core/src/nodes/handled_node_tasks.rs]: Please remove.

**Comment by tomaka** [comment]
Just in general I think it's bad to have style changes in the same PR as something else (or even in general two different changes in the same PR). It's tolerable when there are two or three style changes, but not more. I'm not going to ask you to revert the style changes, but this really pollute the diff.

File: core/src/nodes/handled_node_tasks.rs [hunk 4]
```diff
@@ -450,3 +451,380 @@ where
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    mod task {
```
Comment (tomaka) [inline core/src/nodes/handled_node_tasks.rs]: We shouldn't access anything private in the API. The tests should test the external surface.
I think we don't need these three tests in this module.

File: core/src/nodes/handled_node_tasks.rs [hunk 5]
```diff
@@ -450,3 +451,380 @@ where
...
+                future::ok((peer_id, muxer))
+            };
+            let (node_task, tx, rx) = TestBuilder::new()
+                .with_inner_fut(fut)
+                .with_task_id(345)
+                .node_task();
+
+            // Even though we set up the muxer outbound state to be `Closed` we
+            // still need to open a substream or the outbound state will never
+            // be checked.
```
Comment (tomaka) [inline core/src/nodes/handled_node_tasks.rs]: This should be fixed in the non-test code.

File: core/src/nodes/handled_node_tasks.rs [hunk 6]
```diff
@@ -450,3 +451,380 @@ where
...
+            tx.unbounded_send(create_outbound_substream_event).expect("send msg works");
+            rt.spawn(node_task);
+            let events = rt.block_on(rx.collect()).expect("rx failed");
+
+            assert_eq!(events.len(), 2);
+            assert_matches!(events[0].0, InToExtMessage::NodeReached(PeerId{..}));
+            assert_matches!(events[1].0, InToExtMessage::TaskClosed(Ok(()), _));
+        }
+    }
+    mod handled_node_tasks {
```
Comment (tomaka) [inline core/src/nodes/handled_node_tasks.rs]: Why multiple submodules instead of putting everything in the same?

Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: It was a way to let me run the just the relevant tests while writing them, `cargo t handled_node_tasks` and in the end it seemed like a decent way to group tests. Not to your liking I take it?

File: core/src/nodes/handled_node_tasks.rs (hunk 5 above)
Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: I'll look into it, but I suspect that it's more of an artifact of the `DummyMuxer` than a fault of the actual code.

File: core/src/nodes/handled_node_tasks.rs (hunk 2 above)
Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: > That's not how you do this. The two entries have to be merged into one.

Yeah, it's ugly for sure. Can you show me the better way? I didn't see an obvious way of doing it at the time.

I think it's odd that a struct doesn't include its own name in the debug impl. It was surprising to me when I saw it pop up in the logs and I had to spend time figuring out what it was I was looking for. Principle of least surprise etc.: I think the name should be there.

File: core/src/nodes/handled_node_tasks.rs [hunk 7]
```diff
@@ -450,3 +451,380 @@ where
...
+            dict.insert(id, tx);
+            let task = match dict.entry(id) {
+                Entry::Occupied(inner) => Task{inner},
+                _ => unreachable!()
+            };
+
+            task.close();
+            // REVIEW: this doesn't work because the value is moved. I'd argue
+            // this doesn't need to be tested as it's enforced at compile time.
+            // assert!(!dict.contains_key(&id));
```
Comment (dvdplm) [inline core/src/nodes/handled_node_tasks.rs]: @tomaka do you agree with this? Can I remove this (ignored) test?

**Comment by dvdplm** [comment]
> ust in general I think it's bad to have style changes in the same PR as something else (or even in general two different changes in the same PR).

I agree with this, mixing things makes reviewing much harder than it needs to be. 
I am not sure which style changes you refer to here, but I did run `rustfmt` on two of the test helper files (dummy_transport and dummy_muxer I think) which I think is ok, as the changes there are all relevant to the tests added here. I'll try to pay close attention to it going forward though.

File: core/src/nodes/handled_node_tasks.rs (hunk 5 above)
Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: I am wrong and you're right (and the comment is correct); the `NodeStream` does not check if the outbound state is open or closed (it does check the inbound state).

I'll open an issue and fix in separate PR, ok?

Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: Issue https://github.com/libp2p/rust-libp2p/issues/609

File: core/src/nodes/handled_node_tasks.rs (hunk 2 above)
Comment (tomaka) [reply core/src/nodes/handled_node_tasks.rs]: > I didn't see an obvious way of doing it at the time.

Apologize, looks like there's no built-in way. I probably had memories of another library.

However the fact that there's no built-in way kind of pushes in favour of not doing that. `Vec`, `HashMap`, and all the collections in `std::collections` do the same as this code here, without having their name printed in front.

File: core/src/nodes/handled_node_tasks.rs (hunk 7 above)
Comment (tomaka) [reply core/src/nodes/handled_node_tasks.rs]: Yes

File: core/src/nodes/handled_node_tasks.rs (hunk 2 above)
Comment (dvdplm) [reply core/src/nodes/handled_node_tasks.rs]: This works and does the job:

```
        f.debug_struct("HandledNodesTasks")
            .field("tasks", &self.tasks.keys())
            .field("next_task_id", &self.next_task_id)
            .finish()
```

Yay or nay?

**Comment by dvdplm** [comment]
@tomaka I believe there are two outstanding issues here:

1. Debug impl for `HandledNodesTasks`: do you like the suggested change there?
1. test organization in submodules: do you want me to re-work that?

Other than the above, is there anything else I can do to move this forward?

**Comment by tomaka** [comment]
Yes, please remove the hierarchy of sub-modules in the tests.

I disagree with the surprise factor for changing the Debug implementation. This struct is a collection of tasks, and every single collection in the Rust standard library uses this debug output. To me the reason why `Debug` implementations simply print the struct layout is because programmers are lazy, not because it's the right thing to do.
Let's make @twittner decide?

**Comment by dvdplm** [comment]
> Let's make @twittner decide?

I'll just make the change you prefer and if someone disagrees vehemently we can discuss it again. :)

**Pull request merged**